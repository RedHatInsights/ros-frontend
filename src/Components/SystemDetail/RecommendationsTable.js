import React from 'react';
import { Table, TableHeader, TableBody, expandable } from '@patternfly/react-table';
import propTypes from 'prop-types';
import { flatMap } from 'lodash';
import { EmptyTable } from '@redhat-cloud-services/frontend-components/EmptyTable';
import { EmptyStateDisplay } from '../EmptyStateDisplay/EmptyStateDisplay';
import { CheckCircleIcon } from '@patternfly/react-icons';
import { TextContent, Text, TextVariants } from '@patternfly/react-core';
import './RecommendationsTable.scss';

const renderExpandedView = (row) => {
    return (
        <TextContent className='ros-rec-content'>
            <Text component={TextVariants.p}>
                {row.reason}
            </Text>
            <Text component={TextVariants.p}>
                Recommendations: {row.resolution}
            </Text>
        </TextContent>
    );
};

class RecommendationsTable extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            columns: [{ title: 'Name', cellFormatters: [expandable] }],
            rows: this.createRows()
        };
        this.onCollapse = this.onCollapse.bind(this);
    }

    onCollapse(event, rowKey, isOpen) {
        const { rows } = this.state;
        /**
         * Please do not use rowKey as row index for more complex tables.
         * Rather use some kind of identifier like ID passed with each row.
         */
        rows[rowKey].isOpen = isOpen;
        this.setState({
            rows
        });
    }

    createRows() {
        const rowsData = this.props.recommendations;
        if (rowsData && rowsData.length !== 0) {
            return flatMap(rowsData, (row, index) => {
                return [
                    {
                        id: index,
                        isOpen: index === 0 ? true : false,
                        cells: [{ title: row.description }]
                    },
                    {
                        cells: [{ title: renderExpandedView(row) }],
                        parent: index * 2
                    }
                ];
            });
        } else {
            return [
                {
                    heightAuto: true,
                    cells: [
                        {
                            props: { colSpan: 7 },
                            title: <EmptyTable className='recs-table-empty'>
                                <EmptyStateDisplay title="No recommendations"
                                    text={['No known recommendations affect this system']}
                                    icon={CheckCircleIcon}/>
                            </EmptyTable>
                        }
                    ]
                }
            ];
        }
    }

    render() {
        const { columns, rows } = this.state;
        return (
            <Table aria-label="Expandable table" onCollapse={this.onCollapse}
                variant='compact'
                rows={rows} cells={columns} className="ros-recommendations-table">
                <TableHeader />
                <TableBody />
            </Table>
        );

    }
}

RecommendationsTable.propTypes = {
    recommendations: propTypes.array
};

export default RecommendationsTable;
