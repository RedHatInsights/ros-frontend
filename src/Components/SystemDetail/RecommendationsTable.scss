.ros-recommendations-table.pf-c-table {
  tr {
    --pf-c-table--cell--FontSize: var(--pf-global--FontSize--md);
  }

  .ros-rec-content p {
    margin-bottom: 0;
  }
}