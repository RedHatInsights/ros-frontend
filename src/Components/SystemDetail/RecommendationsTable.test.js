import React from 'react';
import { cleanup } from '@testing-library/react';
import RecommendationsTable from './RecommendationsTable';
import renderer from 'react-test-renderer';

describe('RecommendationsTable component', () => {
    afterEach(cleanup);

    it ('matches snapshot when psi is disabled', ()=>{
        const propValues = {
            recommendations: [{
                ruleId: 'cloud_instance_ros_evaluation|CONSUMPTION_MODEL',
                description: 'This is test rule description.',
                reason: 'This may be caused due to inappropriate consumption model.',
                resolution: 'Please verify this instance.',
                condition: ''
            }],
            psiEnabled: false
        };

        const component = renderer.create(<RecommendationsTable { ...propValues }/>);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it ('matches snapshot when psi is enabled', ()=>{
        const propValues = {
            recommendations: [{
                ruleId: 'cloud_instance_ros_evaluation|CONSUMPTION_MODEL',
                description: 'This is test rule description.',
                reason: 'This may be caused due to inappropriate consumption model.',
                resolution: 'Please verify this instance.',
                condition: ''
            }],
            psiEnabled: true
        };

        const component = renderer.create(<RecommendationsTable { ...propValues }/>);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it ('matches snapshot where there is no suggestions', ()=>{
        const propValues = {
            recommendations: [],
            psiEnabled: false
        };

        const component = renderer.create(<RecommendationsTable { ...propValues }/>);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
