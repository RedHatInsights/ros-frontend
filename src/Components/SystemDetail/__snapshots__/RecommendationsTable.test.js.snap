// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<table
  aria-label="Expandable table"
  className="ros-recommendations-table pf-c-table pf-m-grid-md pf-m-compact"
  data-ouia-component-id="OUIA-Generated-Table-2"
  data-ouia-component-type="PF4/Table"
  data-ouia-safe={true}
  role="grid"
>
  <thead
    className=""
  >
    <tr
      className=""
      data-ouia-component-id="OUIA-Generated-TableRow-1"
      data-ouia-component-type="PF4/TableRow"
      data-ouia-safe={true}
      hidden={false}
    >
      <td
        className=""
        data-key={0}
        data-label=""
        onMouseEnter={[Function]}
        scope=""
      >
        
      </td>
      <th
        className=""
        data-key={1}
        data-label="Name"
        onMouseEnter={[Function]}
        scope="col"
      >
        Name
      </th>
    </tr>
  </thead>
  <tbody
    className="pf-m-expanded"
    role="rowgroup"
  >
    <tr
      className=""
      data-ouia-component-id="OUIA-Generated-TableRow-2"
      data-ouia-component-type="PF4/TableRow"
      data-ouia-safe={true}
      hidden={false}
      onMouseDown={[Function]}
    >
      <td
        className="pf-c-table__toggle"
        data-key={0}
        data-label={null}
        onMouseEnter={[Function]}
      >
        <button
          aria-disabled={false}
          aria-expanded={true}
          aria-label="Details"
          aria-labelledby="simple-node0 expandable-toggle0"
          className="pf-c-button pf-m-plain pf-m-expanded"
          data-ouia-component-id="OUIA-Generated-Button-plain-1"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={false}
          id="expandable-toggle0"
          onClick={[Function]}
          role={null}
          type="button"
        >
          <div
            className="pf-c-table__toggle-icon"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
              />
            </svg>
          </div>
        </button>
      </td>
      <td
        className=""
        data-key={1}
        data-label="Name"
        onMouseEnter={[Function]}
      >
        This is test rule description
      </td>
    </tr>
    <tr
      className="pf-c-table__expandable-row pf-m-expanded"
      data-ouia-component-id="OUIA-Generated-TableRow-3"
      data-ouia-component-type="PF4/TableRow"
      data-ouia-safe={true}
      hidden={false}
      onMouseDown={[Function]}
    >
      <td
        className=""
        data-key={0}
        data-label={null}
        onMouseEnter={[Function]}
      />
      <td
        className=""
        colSpan={1}
        data-key={1}
        data-label="Name"
        id="expanded-content1-1"
        onMouseEnter={[Function]}
      >
        <div
          className="pf-c-table__expandable-row-content"
        >
          <div
            className="pf-c-content"
          >
            <p
              className=""
              data-pf-content={true}
            >
              <p
                className=""
                data-pf-content={true}
              >
                <strong>
                  Detected issues
                </strong>
              </p>
              This may be caused due to inappropriate consumption model
            </p>
            <p
              className=""
              data-pf-content={true}
            >
              <p
                className=""
                data-pf-content={true}
              >
                <strong>
                  Recommendation
                </strong>
              </p>
              Please verify this instance
            </p>
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
